Adding the `drakeSimulator` class and a working example of `drake_walking.ipynb`

akhilsathuluri opened the pull request:
This is a draft PR for the newly added class which are bindings of `CoMoDo` with `drake` and an example showing how it works.
The class follows almost exactly the same design of the `mujocoSimulator` class.
I also added a helper class that bridges between any one coming from a different platform to make their URDFs compatible with drake and simplifies loading. However, this needs the `odio_urdf` from my repo: https://github.com/akhilsathuluri/odio_urdf and some additional dependencies, which needs to be handled seperately. 

For the sake of a common structure, I added the example in Jupyter too, but it is not comfortable to maintain it (especially when the JSON fails due to git merge issues). An alternative is: https://github.com/mwouts/jupytext

In the next push, I will have modified `include` files to add the `angular_momentum` and `contact_force_regularisation` tasks sclaed by robots mass. 
(I changed the includes for faster prototyping.)
I have observed the same things as you mentioned today, that the robot falls down after a couple of steps in `mujoco` too.

traversaro commented:
> However, this needs the `odio_urdf` from my repo: https://github.com/akhilsathuluri/odio_urdf and some additional dependencies, which needs to be handled seperately.

As `odio_urdf` mantainer are not replying, could we do a friendly fork of it called `amo_urdf` (the italian joke, "odio" means "I hate", "amo" I love) so that we can depend on it without problems?

akhilsathuluri commented:
> > However, this needs the `odio_urdf` from my repo: https://github.com/akhilsathuluri/odio_urdf and some additional dependencies, which needs to be handled seperately.
> 
> As `odio_urdf` mantainer are not replying, could we do a friendly fork of it called `amo_urdf` (the italian joke, "odio" means "I hate", "amo" I love) so that we can depend on it without problems?

@traversaro, please let me know what to do about this.

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+            self.active_meshcat = True
+        pass
+
+    # does this need to simulator specific?
```
Could you remove this comment and open an issue related to this point? I like the idea but it should be in an issue not to loose track of it

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+        pass
+
+    # does this need to simulator specific?
+    def set_base_pose_in_drake(self, xyz_rpy):
```
why `in_drake` ? We should port this method abstract in the base class, however I do not see the point to call it in drake since it is a method of the drake class

CarlottaSartore replied:
I think we can manage this in a separate PR

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+        self.qpos[:7] = base_xyz_quat
+
+    # set initial positions of the joints
+    def set_joint_vector_in_drake(self, pos):
```
same for this name, Why the comment says it specific for the initial position ?

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+        pass
+
+    def step(self, n_step=1, visualize=True):
+        # TODO: does nothing currently
```
Could you remove this comment? Since now is implemented

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+            self.diagram.ForcedPublish(self.context)
+        pass
+
+    def step_with_motors(self, n_step, torque):
```
Could you print a warning that says that the method is not implemented ?

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+        # order -- linar_vel + angular_vel
+        return np.concatenate((base_vel[3:], base_vel[:3]))
+
+    # TODO: Change from get_state to get_joint_state?
```
Could you remove this and open an issue please ?

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+            self.joint_torques,
+        )
+
+    def close(self):
```
What should be this issue ? could you either document (with a warning) or remove it please ?

CarlottaSartore replied:
I thin this is now resolved

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+
+    def get_simulation_frequency(self):
+        # return the simulator timestep
+        return self.time_step
```
Why timestep has the frequency ?

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+        qy = cr * sp * cy + sr * cp * sy
+        qz = cr * cp * sy - sr * sp * cy
+
+        # Note: The order for drake is different as compared to mujoco
```
How ? could you document this differences in an issue ?

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+            the visualisation."
+        )
+        self.reset()
+        pass
```
why there is a pass ?

CarlottaSartore commented on src/comodo/drakeSimulator/utils.py:
```diff
@@ -0,0 +1,209 @@
+        AddContactMaterial(
+            1e4, 1e7, surface_friction_ground, proximity_properties_ground
+        )
+        AddRigidHydroelasticProperties(0.01, proximity_properties_ground)
```
why these values are hardcoded?

CarlottaSartore replied:
Could you please add a comment (also in this PR is fine) about this value ? thanks

CarlottaSartore commented on src/comodo/drakeSimulator/utils.py:
```diff
@@ -0,0 +1,209 @@
+        proximity_properties_feet = ProximityProperties()
+        AddContactMaterial(1e4, 1e7, surface_friction_feet, proximity_properties_feet)
+        AddCompliantHydroelasticProperties(0.01, 1e8, proximity_properties_feet)
+        for ii in ["l_foot_front", "l_foot_rear", "r_foot_front", "r_foot_rear"]:
```
this frame name are hardcoded , it would be therefore impossible to use with a different robot model. The robot model class have info about the contact frame name, you can add over there the names you need (not hardcoded but set )

CarlottaSartore replied:
Resolving it

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+            file_type="urdf",
+        )[0]
+
+        # TODO: Use Im and Km to add motor params here
```
Could you add a warning that the motor parameters are not utilized ?

CarlottaSartore replied:
Done  here 

```
logging.warning("NotImplementedWarning: Motor parameters are currently ignored.")
```
resolving

CarlottaSartore commented on src/comodo/drakeSimulator/utils.py:
```diff
@@ -0,0 +1,209 @@
+            # replace the urdf to load .obj files
+            child.set("filename", path.replace("stl", "obj"))
+
+        # # if the same, will be ignored else converted
```
Could you remove  these comments and open an issue if is an issue needed to be addressed

CarlottaSartore commented on src/comodo/drakeSimulator/utils.py:
```diff
@@ -0,0 +1,209 @@
+            RigidTransform(),
+            HalfSpace(),
+            "ground_visual",
+            np.array([0.5, 0.5, 0.5, 0.0]),
```
what is this value ?

CarlottaSartore replied:
Could you please add a comment about this value ? Thanks

CarlottaSartore replied:
Done in b53d0091f81dc73c27952c6bdcb393d72d846767

CarlottaSartore commented on src/comodo/drakeSimulator/utils.py:
```diff
@@ -0,0 +1,209 @@
+                # RigidTransform(),
+                BoxDrake((xMinMax[1] - xMinMax[0]) / 2, yMinMax[1] - yMinMax[0], 2e-3),
+                ii + "_collision",
+                np.array([1.0, 1.0, 1.0, 1]),
```
What is this value ?

CarlottaSartore replied:
done in b53d0091f81dc73c27952c6bdcb393d72d846767 ( they are RGBA values for the feet )

CarlottaSartore reviewed (changes requested):
I tested also on my machine and everything works fine !  Thank you really much @akhilsathuluri  for the PR ! 
Few comments needs to be addressed before merging, I would also ask you if you can generally clean up the code from comments that do not document the code, and open issues if there is anything you would like to implement in the future or features you would like to have in comodo ! Thanks

akhilsathuluri commented on README.md:
```diff
@@ -42,6 +42,17 @@ pip install --no-deps -e .
 ## Usage 
```
Yes, the examples and the general class execution needs the added installations. I will move it to the general installation information.

akhilsathuluri commented:
@CarlottaSartore : I modified some of the methods based on the comments. Do have a look and let me know.

CarlottaSartore commented on examples/drake_walking.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Comodo Exmaple \n",
```
Done

CarlottaSartore commented on examples/drake_walking.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "\n",
+    "    if counter == n_step_mpc_tsid:\n",
+    "        counter = 0\n",
+    "if  drake_instance.visualize_robot_flag:\n",
```
I think we can adress this in a separate PR

akhilsathuluri commented:
Hello @CarlottaSartore, here are the changes:
- removed hardcoded feet collisions (x and y values for the boxes)
- added argument names for collision properties (foot/ground dissipation and stiffness)
- Added comments for other hardcoded numbers (RGBA values for visual and collision meshes for the ground plane)

Please let me know if you think more changes are necessary. Also, apologies for the delay 😅

CarlottaSartore commented on src/comodo/drakeSimulator/drakeSimulator.py:
```diff
@@ -0,0 +1,243 @@
+        self.duh.add_ground_with_friction(plant)
+
+        # configure feet collisions
+        xMinMax = [-0.1, 0.1]
```
Done in b53d0091f81dc73c27952c6bdcb393d72d846767

CarlottaSartore commented:
Everything looks in place, thanks for your contribution @akhilsathuluri  :)
